ext/opcache/Optimizer/ssa_integrity.c

```diff
@@ -0,0 +1,405 @@
+						VAR(ssa_op->result_def), INSTR(i));
+			}
+			if (VAR_NUM(opline->result.var) != ssa->vars[ssa_op->result_def].var) {
+				FAIL("result def of " VARFMT " does not match op %d of " INSTRFMT "\n",
+						VAR(ssa_op->result_def), VAR_NUM(opline->result.var), INSTR(i));
+			}
+		}
+	} FOREACH_INSTR_NUM_END();
+
+	/* Phis */
+	FOREACH_PHI(phi) {
+		int source;
+		FOREACH_PHI_SOURCE(phi, source) {
+			if (source < 0) {
+				FAIL(VARFMT " negative source\n", VAR(phi->ssa_var));
+			}
+			if (!is_in_phi_use_chain(ssa, source, phi)) {
+				FAIL(VARFMT " not in phi use chain of %d\n", VAR(phi->ssa_var), source);
+			}
+			if (ssa->vars[source].var != ssa->vars[phi->ssa_var].var) {
+				FAIL(VARFMT " source of phi for " VARFMT "\n", VAR(source), VAR(phi->ssa_var));
+			}
+		} FOREACH_PHI_SOURCE_END();
+		if (ssa->vars[phi->ssa_var].definition_phi != phi) {
+			FAIL(VARFMT " does not define this phi\n", VAR(phi->ssa_var));
+		}
+	} FOREACH_PHI_END();
+
+	/* Blocks */
+	for (i = 0; i < cfg->blocks_count; i++) {
+		zend_basic_block *block = &cfg->blocks[i];
+		int *predecessors = &cfg->predecessors[block->predecessor_offset];
+		int s, j;
+
+		if (i != 0 && block->start < (block-1)->start + (block-1)->len) {
+			FAIL("Block %d start %d smaller previous end %d\n",
+				i, block->start, (block-1)->start + (block-1)->len);
+		}
+		if (i != cfg->blocks_count-1 && block->start + block->len > (block+1)->start) {
+			FAIL("Block %d end %d greater next start %d\n",
+				i, block->start + block->len, (block+1)->start);
+		}
+
+		for (j = block->start; j < block->start + block->len; j++) {
+			if (cfg->map[j] != i) {
+				FAIL("Instr " INSTRFMT " not associated with block %d\n", INSTR(j), i);
+			}
+		}
+
+		if (!(block->flags & ZEND_BB_REACHABLE)) {
+			if (ssa->blocks[i].phis) {
+				FAIL("Unreachable block %d has phis\n", i);
+			}
+			continue;
+		}
+
+		for (s = 0; s < block->successors_count; s++) {
+			zend_basic_block *next_block;
+			if (block->successors[s] < 0) {
+				FAIL("Successor number %d of %d negative", s, i);
+			}
+			next_block = &cfg->blocks[block->successors[s]];
+			if (!(next_block->flags & ZEND_BB_REACHABLE)) {
+				FAIL("Successor %d of %d not reachable\n", block->successors[s], i);
+			}
+			if (!is_in_predecessors(cfg, next_block, i)) {
+				FAIL("Block %d predecessors missing %d\n", block->successors[s], i);
+			}
+		}
+
+		for (j = 0; j < block->predecessors_count; j++) {
+			if (predecessors[j] >= 0) {
+				zend_basic_block *prev_block = &cfg->blocks[predecessors[j]];
+				if (!(prev_block->flags & ZEND_BB_REACHABLE)) {
+					FAIL("Predecessor %d of %d not reachable\n", predecessors[j], i);
+				}
+				if (!is_in_successors(prev_block, i)) {
+					FAIL("Block %d successors missing %d\n", predecessors[j], i);
+				}
+				// TODO Right now CFG construction can generate the same predecessor multiple
+				// times. Decide whether or not it should do that, for now disable the check.
```

Comment:
With the change in zend_cfg.c to avoid multiple identical predecessors, it might be possible to reenable this code.